f/frizzle/comapeo/comapeo_observations.py

```diff
@@ -0,0 +1,356 @@
+
+            # Process tags
+            if "tags" in observation:
+                for key, value in observation["tags"].items():
+                    new_key = key.replace("-", "_")
+                    observation[new_key] = value
+                del observation["tags"]
+
+            # Process attachments
+            if "attachments" in observation:
+                filenames = []
+                for attachment in observation["attachments"]:
+                    if "url" in attachment:
+                        logger.info(attachment["url"])
+                        response = requests.get(attachment["url"], headers=headers)
+                        if response.status_code == 200:
+                            content_type = response.headers.get("Content-Type", "")
+                            extension = mimetypes.guess_extension(content_type) or ""
+
+                            file_name = os.path.basename(attachment["url"]) + extension
+                            filenames.append(file_name)
+
+                            save_path = os.path.join(
+                                attachment_root,
+                                "comapeo",
+                                sanitized_project_name,
+                                "attachments",
+                                file_name,
+                            )
+                            os.makedirs(os.path.dirname(save_path), exist_ok=True)
+                            with open(save_path, "wb") as file:
+                                file.write(response.content)
+                            logger.info("Download completed.")
+                        else:
+                            # TODO: add retries
+                            try:
+                                error_message = response.json().get(
+                                    "message", "Unknown error"
+                                )
+                            except requests.exceptions.JSONDecodeError:
+                                error_message = "No JSON response received."
+                            logger.error(
+                                f"Failed downloading attachments. Error: {error_message}"
+                            )
+
+                observation["attachments"] = ", ".join(filenames)
+
+        # Store observations in a dictionary with project_id as key
+        comapeo_data[sanitized_project_name] = current_project_data
+
+        logger.info(
+            f"Project {index + 1} (ID: {project_id}, name: {project_name}): Processed {len(current_project_data)} observation(s)."
+        )
+    return comapeo_data
+
+
+def camel_to_snake(name):
+    return re.sub("([a-z0-9])([A-Z])", r"\1_\2", name).lower()
+
+
+class CoMapeoDBWriter:
+    """
+    Converts unstructured CoMapeo observations data to structured SQL tables.
+
+    Methods:
+        __init__(db_connection_string, table_name): Initializes the CoMapeoDBWriter with the provided connection string and project to be used.
```

Comment:
Can remove this "Methods: __init__" couple of lines. It's already in a better place on line 207

Reply:
Good catch! Thanks.